File: backend/app/schemas/activities.py
"""Schémas Pydantic v2 — Activités."""
from __future__ import annotations

import uuid
from datetime import datetime
from typing import Literal

from pydantic import BaseModel, Field

ActivityType = Literal["Appel", "Email", "Tâche", "RDV"]
RelatedType = Literal["contact", "deal", "project"]


class ActivityCreate(BaseModel):
    type: ActivityType
    when: datetime
    duration_min: int | None = Field(default=None, ge=1, le=1440)
    outcome: str | None = Field(default=None, max_length=100)
    notes: str | None = Field(default=None, max_length=10000)
    related_type: RelatedType | None = None
    related_id: uuid.UUID | None = None
    reminder_at: datetime | None = None


class ActivityPatch(BaseModel):
    type: ActivityType | None = None
    when: datetime | None = None
    duration_min: int | None = Field(default=None, ge=1, le=1440)
    outcome: str | None = Field(default=None, max_length=100)
    notes: str | None = Field(default=None, max_length=10000)
    related_type: RelatedType | None = None
    related_id: uuid.UUID | None = None
    reminder_at: datetime | None = None
    reminder_sent: bool | None = None


class ActivityOut(BaseModel):
    id: uuid.UUID
    type: ActivityType
    when: datetime
    duration_min: int | None
    outcome: str | None
    notes: str | None
    related_type: str | None
    related_id: uuid.UUID | None
    reminder_at: datetime | None
    reminder_sent: bool
    user_id: uuid.UUID | None
    created_at: datetime
    updated_at: datetime
    # Champs enrichis
    related_label: str | None = None  # ex. "Alice Martin" pour un contact

    model_config = {"from_attributes": True}


class ActivityList(BaseModel):
    items: list[ActivityOut]
    total: int

File: frontend/src/api/activities.ts
import { useMutation, useQuery, useQueryClient } from '@tanstack/react-query'
import { apiClient } from './client'
import { useAuthStore } from '@/store/authStore'

// ── Types ─────────────────────────────────────────────────────────────────────

export type ActivityType = 'Appel' | 'Email' | 'Tâche' | 'RDV'
export type ActivityRelatedType = 'contact' | 'deal' | 'project'

export interface ActivityOut {
    id: string
    type: ActivityType
    when: string
    duration_min: number | null
    outcome: string | null
    notes: string | null
    related_type: ActivityRelatedType | null
    related_id: string | null
    related_label: string | null
    reminder_at: string | null
    reminder_sent: boolean
    user_id: string | null
    created_at: string
    updated_at: string
}

export interface ActivityList {
    items: ActivityOut[]
    total: number
}

export interface ActivityCreate {
    type: ActivityType
    when: string
    duration_min?: number | null
    outcome?: string | null
    notes?: string | null
    related_type?: ActivityRelatedType | null
    related_id?: string | null
    reminder_at?: string | null
}

export interface ActivityPatch {
    type?: ActivityType
    when?: string
    duration_min?: number | null
    outcome?: string | null
    notes?: string | null
    related_type?: ActivityRelatedType | null
    related_id?: string | null
    reminder_at?: string | null
    reminder_sent?: boolean
}

export interface ActivitiesParams {
    related_type?: ActivityRelatedType
    related_id?: string
    type?: ActivityType
    date_from?: string
    date_to?: string
    page?: number
    page_size?: number
}

// ── Keys ─────────────────────────────────────────────────────────────────────

export const activityKeys = {
    all: ['activities'] as const,
    list: (params: ActivitiesParams) => [...activityKeys.all, 'list', params] as const,
    upcoming: () => [...activityKeys.all, 'upcoming'] as const,
    detail: (id: string) => [...activityKeys.all, 'detail', id] as const,
}

// ── Queries ───────────────────────────────────────────────────────────────────

export function useActivities(params: ActivitiesParams = {}) {
    return useQuery({
        queryKey: activityKeys.list(params),
        queryFn: () =>
            apiClient
                .get<ActivityList>('/activities', {
                    params: Object.fromEntries(
                        Object.entries(params).filter(([, v]) => v !== undefined && v !== null),
                    ),
                })
                .then((r) => r.data),
    })
}

export function useUpcomingActivities(hours = 48) {
    const accessToken = useAuthStore((s) => s.accessToken)
    return useQuery({
        queryKey: activityKeys.upcoming(),
        enabled: !!accessToken,
        staleTime: 60_000, // 1 minute — évite les requêtes répétées sur focus
        queryFn: () =>
            apiClient
                .get<ActivityList>('/activities/upcoming', { params: { hours } })
                .then((r) => r.data),
    })
}

export function useActivity(id: string | undefined) {
    return useQuery({
        queryKey: activityKeys.detail(id!),
        queryFn: () => apiClient.get<ActivityOut>(`/activities/${id}`).then((r) => r.data),
        enabled: !!id,
    })
}

// ── Mutations ─────────────────────────────────────────────────────────────────

export function useCreateActivity() {
    const qc = useQueryClient()
    return useMutation({
        mutationFn: (data: ActivityCreate) =>
            apiClient.post<ActivityOut>('/activities', data).then((r) => r.data),
        onSuccess: () => qc.invalidateQueries({ queryKey: activityKeys.all }),
    })
}

export function usePatchActivity() {
    const qc = useQueryClient()
    return useMutation({
        mutationFn: ({ id, data }: { id: string; data: ActivityPatch }) =>
            apiClient.patch<ActivityOut>(`/activities/${id}`, data).then((r) => r.data),
        onSuccess: () => qc.invalidateQueries({ queryKey: activityKeys.all }),
    })
}

export function useDeleteActivity() {
    const qc = useQueryClient()
    return useMutation({
        mutationFn: (id: string) => apiClient.delete(`/activities/${id}`),
        onSuccess: () => qc.invalidateQueries({ queryKey: activityKeys.all }),
    })
}
